service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/vm/CreateAzureVmStep.java

```diff
@@ -177,9 +177,22 @@ public StepResult undoStep(FlightContext context) throws InterruptedException {
     ComputeManager computeManager = crlService.getComputeManager(azureCloudContext, azureConfig);
 
     try {
+      VirtualMachine resolvedVm =
+              computeManager
+                      .virtualMachines()
+                      .getByResourceGroup(
+                              azureCloudContext.getAzureResourceGroupId(), resource.getVmName());
+
       computeManager
-          .virtualMachines()
-          .deleteByResourceGroup(azureCloudContext.getAzureResourceGroupId(), resource.getVmName());
+              .virtualMachines()
+              .deleteByResourceGroup(azureCloudContext.getAzureResourceGroupId(), resource.getVmName());
+
+      resolvedVm
+              .networkInterfaceIds()
+              .forEach(nic -> computeManager.networkManager().networkInterfaces().deleteById(nic));
+
+      // Delete the OS disk
+      computeManager.disks().deleteById(resolvedVm.osDiskId());
```

Comment:
I wonder if you'll need to keep coming back here to add things to be deleted. That would be easy to miss, so I wonder if you wouldn't be better served by a central `deleteVmAndEverythingInIt()` method, especially if you need to do this in another place, like a delete endpoint.

Reply:
where would you but this central method

Reply:
Wherever your delete endpoint is, or a utility class. If you only ever need it here, then there's no need to break it out.